File: 백준/Gold/2268. 수들의 합 7/수들의 합 7.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    static int n, q, a, b, c;

    static int[] data;
    static long[] tree;

    static long init(int idx, int start, int end) {
        if (start == end) return tree[idx] = data[start];
        int mid = (start + end) >> 1;
        return tree[idx] = init(idx << 1, start, mid) + init((idx << 1) + 1, mid + 1, end);
    }

    static long sum(int start, int end, int idx, int left, int right) {
        if (left > end || right < start) return 0;
        if (left <= start && end <= right) return tree[idx];
        int mid = (start + end) >> 1;
        return sum(start, mid, idx << 1, left, right) + sum(mid + 1, end, (idx << 1) + 1, left, right);
    }

    static long update(int start, int end, int node, int idx, long tmp) {
        if (idx < start || idx > end) return tree[node];
        if (start == end) return tree[node] = tmp;
        int mid = (start + end) >> 1;
        return tree[node] = update(start, mid, node << 1, idx, tmp) + update(mid + 1, end, (node << 1) + 1, idx, tmp);
    }

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        q = Integer.parseInt(st.nextToken());

        data = new int[n + 1];
        tree = new long[n * 4];

        init(1, 1, n);

        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());

            int max = Math.max(b, c);
            int min = Math.min(b, c);

            if (a == 0) bw.write(sum(1, n, 1, min, max) + "\n");
            else update(1, n, 1, b, c);
        }
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/2268. 수들의 합 7/수들의 합 7.py
import sys

input = lambda: sys.stdin.readline().rstrip()


class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.tree = [0] * (n + 1)

    def update(self, idx, delta):
        while idx <= self.n:
            self.tree[idx] += delta
            idx += idx & -idx

    def prefix_sum(self, idx):
        result = 0
        while idx > 0:
            result += self.tree[idx]
            idx -= idx & -idx
        return result

    def range_sum(self, left, right):
        return self.prefix_sum(right) - self.prefix_sum(left - 1)


N, M = map(int, input().split())
fenwick = FenwickTree(N)
A = [0] * (N + 1)

for _ in range(M):
    command = list(map(int, input().split()))
    if command[0] == 1:
        # Modify(i, k): A[i] = k
        i, k = command[1], command[2]
        delta = k - A[i]
        A[i] = k
        fenwick.update(i, delta)
    else:
        # Sum(i, j): 구간 합을 계산
        i, j = command[1], command[2]
        if i > j:
            i, j = j, i
        print(fenwick.range_sum(i, j))